ethcontract/src/tokens.rs

```diff
@@ -0,0 +1,532 @@
+// that is implemented for all types that implement `SingleTokenize` except `Vec<u8>` and
+// `[u8; n]` and then only implement `SingleTokenize` for vectors and arrays of
+// `SingleTokenizeArray`.
+//
+// The drawback is that if a solidity function actually used an array of u8 instead of bytes then we
+// would not be able to interact with it. An alternative solution could be to use a Bytes new type
+// but this makes calling those functions slightly more annoying.
+//
+// In some cases like when passing arguments to `MethodBuilder` or decoding events we need to be
+// able to pack multiple types into a single generic parameter. This is accomplished by
+// `MultiTokenize` and `MultiDetokenize` which work on vectors of tokens instead of a single token.
+// This is split into two traits instead of one like for `SingleTokenize` to avoid more work when
+// only one half of the functionality is needed. For example events are only ever decoded so there
+// is no reason to implement encoding.
+//
+// A completely different approach could be to avoid using the trait system and instead encode all
+// rust types into tokens directly in the ethcontract generated bindings.
+
+use crate::I256;
+use arrayvec::ArrayVec;
+use ethcontract_common::{abi::Token, TransactionHash};
+use std::convert::TryInto;
+use web3::types::{Address, U256};
+
+/// A tokenization related error.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// SingleTokenize::from_token token type doesn't match the rust type.
+    #[error("expected a different token type")]
+    TypeMismatch,
+    /// SingleTokenize::from_token is called with integer that doesn't fit in the rust type.
+    #[error("abi integer is does not fit rust integer")]
+    IntegerMismatch,
+    /// SingleTokenize::from_token token is fixed bytes with wrong length.
+    #[error("expected a different number of fixed bytes")]
+    FixedBytesLengthsMismatch,
+    /// SingleTokenize::from_token token is fixed array with wrong length.
+    #[error("expected a different number of tokens in fixed array")]
+    FixedArrayLengthsMismatch,
+    /// SingleTokenize::from_token token is tuple with wrong length.
+    #[error("expected a different number of tokens in tuple")]
+    TupleLengthMismatch,
+    /// MultiDetokenize::from_tokens has wrong number of tokens.
+    #[error("expected a different number of tokens when detokenizing multiple tokens")]
+    MultiDetokenizeLengthMismatch,
+}
+
+/// Rust type and single token conversion.
+pub trait SingleTokenize {
+    /// Convert self into token.
+    fn from_token(token: Token) -> Result<Self, Error>
+    where
+        Self: Sized;
+
+    /// Convert token into Self.
+    fn into_token(self) -> Token;
+}
+
+impl SingleTokenize for Vec<u8> {
+    fn from_token(token: Token) -> Result<Self, Error>
+    where
+        Self: Sized,
+    {
+        match token {
+            Token::Bytes(bytes) => Ok(bytes),
```

Comment:
Isn't this missing the FixedBytes arm?

Reply:
This is intentional because we always translate Token::Bytes to Vec<u8> so it should never be the case that we call this with FixedBytes. This is an example where we can be stricter than the rust-web3 version which presumably wants to give users more flexibility by allowing either version of bytes.